File: src/main/java/de/sean/splugin/spigot/events/HopperEvent.java
package de.sean.splugin.spigot.events;

/* SPlugin */
import de.sean.splugin.util.SLockUtil;

/* Spigot */
import org.bukkit.Location;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryMoveItemEvent;
import org.bukkit.event.inventory.InventoryType;

/* NBT-API */
import de.tr7zw.nbtapi.NBTTileEntity;

public class HopperEvent implements Listener {
    @EventHandler
    public void onItemMove(final InventoryMoveItemEvent event) {
        if (event.getDestination().getType() == InventoryType.HOPPER) {
            // This is a hopper trying to pull from something.
            switch (event.getSource().getType()) {
                case CHEST:
                case FURNACE:
                case BARREL:
                case SHULKER_BOX:
                    // This hopper is trying to pull from some inventory which *may* be locked.
                    // Note: we do not have to check for double chests, as both sides of a chest are individually locked.
                    final Location sourceLocation = event.getSource().getLocation();
                    if (sourceLocation == null) break;
                    final NBTTileEntity source = new NBTTileEntity(sourceLocation.getBlock().getState());
                    if (!source.getBoolean(SLockUtil.REDSTONE_ATTRIBUTE)) event.setCancelled(true);
                    break;
            }
        }
    }
}

File: src/main/kotlin/de/sean/splugin/spigot/events/HopperEvent.kt
package de.sean.splugin.spigot.events

import de.sean.splugin.util.SLockUtil
import de.tr7zw.nbtapi.NBTTileEntity
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryMoveItemEvent
import org.bukkit.event.inventory.InventoryType

class HopperEvent : Listener {
    @EventHandler
    fun onItemMove(event: InventoryMoveItemEvent) {
        if (event.destination.type == InventoryType.HOPPER) {
            // This is a hopper trying to pull from something.
            when (event.source.type) {
                InventoryType.CHEST, InventoryType.FURNACE, InventoryType.BARREL, InventoryType.SHULKER_BOX -> {
                    // This hopper is trying to pull from some inventory which *may* be locked.
                    // Note: we do not have to check for double chests, as both sides of a chest are individually locked.
                    val sourceLocation = event.source.location ?: return
                    val source = NBTTileEntity(sourceLocation.block.state)
                    if (!source.getBoolean(SLockUtil.REDSTONE_ATTRIBUTE)) event.isCancelled = true
                }
                else -> return
            }
        }
    }
}
